File: metacasa-ios/MetaCasa/Models/Goal.swift
import Foundation

struct Goal: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var householdId: UUID
    var name: String
    var description: String?
    var targetAmount: Decimal
    var currentAmount: Decimal
    var currency: String
    var targetDate: Date?
    var status: GoalStatus
    var icon: String?
    var color: String?
    var priority: Int
    var category: String?
    var accountId: UUID?
    var notes: String?
    let createdBy: UUID
    let createdAt: Date?
    var updatedAt: Date?
    var completedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case householdId = "household_id"
        case name
        case description
        case targetAmount = "target_amount"
        case currentAmount = "current_amount"
        case currency
        case targetDate = "target_date"
        case status
        case icon
        case color
        case priority
        case category
        case accountId = "account_id"
        case notes
        case createdBy = "created_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case completedAt = "completed_at"
    }

    var progress: Double {
        guard targetAmount > 0 else { return 0 }
        return min(1, (currentAmount as NSDecimalNumber).doubleValue / (targetAmount as NSDecimalNumber).doubleValue)
    }
}

enum GoalStatus: String, Codable, Hashable, Sendable, CaseIterable {
    case active, completed, paused, canceled

    var label: String {
        switch self {
        case .active: "Activa"
        case .completed: "Completada"
        case .paused: "Pausada"
        case .canceled: "Cancelada"
        }
    }
}

struct GoalContribution: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    let goalId: UUID
    var amount: Decimal
    let contributedBy: UUID
    var contributedAt: Date
    var notes: String?
    var transactionId: UUID?

    enum CodingKeys: String, CodingKey {
        case id
        case goalId = "goal_id"
        case amount
        case contributedBy = "contributed_by"
        case contributedAt = "contributed_at"
        case notes
        case transactionId = "transaction_id"
    }
}

File: metacasa-flutter/lib/models/goal.dart
import 'package:decimal/decimal.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../core/utils/decimal_converter.dart';

part 'goal.freezed.dart';
part 'goal.g.dart';

/// Estado de la meta de ahorro.
enum GoalStatus {
  @JsonValue('active')
  active,
  @JsonValue('completed')
  completed,
  @JsonValue('paused')
  paused,
  @JsonValue('canceled')
  canceled;

  String get label => switch (this) {
        GoalStatus.active => 'Activa',
        GoalStatus.completed => 'Completada',
        GoalStatus.paused => 'Pausada',
        GoalStatus.canceled => 'Cancelada',
      };
}

/// Meta de ahorro. Espejo 1:1 de `public.goals`.
@freezed
abstract class Goal with _$Goal {
  const Goal._();

  const factory Goal({
    required String id,
    @JsonKey(name: 'household_id') required String householdId,
    required String name,
    String? description,
    @JsonKey(name: 'target_amount')
    @DecimalConverter()
    required Decimal targetAmount,
    @JsonKey(name: 'current_amount')
    @DecimalConverter()
    required Decimal currentAmount,
    required String currency,
    @JsonKey(name: 'target_date') DateTime? targetDate,
    required GoalStatus status,
    String? icon,
    String? color,
    required int priority,
    String? category,
    @JsonKey(name: 'account_id') String? accountId,
    String? notes,
    @JsonKey(name: 'created_by') required String createdBy,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
    @JsonKey(name: 'completed_at') DateTime? completedAt,
  }) = _Goal;

  factory Goal.fromJson(Map<String, dynamic> json) => _$GoalFromJson(json);

  /// Progreso hacia la meta (0.0 – 1.0).
  double get progress {
    if (targetAmount <= Decimal.zero) return 0;
    final double ratio = currentAmount.toDouble() / targetAmount.toDouble();
    return ratio < 1 ? ratio : 1;
  }
}

/// Aporte a una meta. Espejo 1:1 de `public.goal_contributions`.
@freezed
abstract class GoalContribution with _$GoalContribution {
  const factory GoalContribution({
    required String id,
    @JsonKey(name: 'goal_id') required String goalId,
    @DecimalConverter() required Decimal amount,
    @JsonKey(name: 'contributed_by') required String contributedBy,
    @JsonKey(name: 'contributed_at') required DateTime contributedAt,
    String? notes,
    @JsonKey(name: 'transaction_id') String? transactionId,
  }) = _GoalContribution;

  factory GoalContribution.fromJson(Map<String, dynamic> json) =>
      _$GoalContributionFromJson(json);
}
